### ISSUE TITLE
Producer blocks when broker is unreachable

### ISSUE THREAD

**Issue by Jox4** [Parsely/pykafka#498]
```python
logger.info('Sending events...')
for sip_event in self.events_list:
     try:
         self.sip_producer.produce(msg)
     except:
         while True:
             logger.error('Could not connect to Kafka')
             logger.error(traceback.format_exc())
             self.reconnect()
             try:
                 self.sip_producer.produce(msg)
                 break
             except:
                 pass
```
This code is blocked when server is disconected. Don't catch timeout or any exception. With 2.1.1 is running well.

**PyKafka version**: 2.2.1
**Kafka version**: 2.9.1

**Comment by emmett9001** [comment]
Hi @Jox4, thanks for the ticket. Can you be more specific about the behavior you're seeing? How are you creating your `Producer` instance, and what keyword arguments are you using? What's the replication factor of your topic, how many brokers are you running, and how many of those brokers are going down during your test?

**Comment by Jox4** [comment]
Client instance:
`self.client = StoneKafka(
            hosts="server_ip:server_port",
            socket_timeout_ms=10 * 1000,
            offsets_channel_socket_timeout_ms=8 * 1000,
        )`

It's a simply test of our application: 1 broker, 1 replication factor and 128 partitions (for the balanced consumers).


Thanks for your attention

**Comment by emmett9001** [comment]
We've seen this same issue happen at Parsely, so we may be able to dig up some more server logs that will help us fix it.

**Comment by emmett9001** [comment]
@Jox4 I'm unable to reliably replicate the hang you're mentioning, though I believe I've come across it once. Are you opposed to using delivery reports? This is the standard method that pykafka provides to verify the receipt of produced messages. Maybe there's some reason you're unable to use them?

**Comment by emmett9001** [comment]
Potentially related traceback from a Parse.ly test machine
```
2016-07-12 20:57:07,316 - pykafka.producer - DEBUG - Sending 944 messages to broker 12
2016-07-12 20:57:07,801 - pykafka.producer - WARNING - Broker ue1d-kafka-beta1a.lan.cogtree.com:9092 disconnected. Retrying.
2016-07-12 20:57:07,801 [WARNING] pykafka.producer - Broker ue1d-kafka-beta1a.lan.cogtree.com:9092 disconnected. Retrying.
2016-07-12 20:57:07,801 - pykafka.cluster - DEBUG - Updating cluster, attempt 1/3
2016-07-12 20:57:07,802 - pykafka.connection - DEBUG - Connecting to ue1d-kafka-beta1a.lan.cogtree.com:9092
2016-07-12 20:57:07,802 - pykafka.producer - WARNING - Broker ue1d-kafka-beta1b.lan.cogtree.com:9092 disconnected. Retrying.
2016-07-12 20:57:07,802 [WARNING] pykafka.producer - Broker ue1d-kafka-beta1b.lan.cogtree.com:9092 disconnected. Retrying.
2016-07-12 20:57:07,804 - pykafka.producer - WARNING - Broker ue1d-kafka-beta1c.lan.cogtree.com:9092 disconnected. Retrying.
2016-07-12 20:57:07,804 [WARNING] pykafka.producer - Broker ue1d-kafka-beta1c.lan.cogtree.com:9092 disconnected. Retrying.
2016-07-12 20:57:07,812 - pykafka.cluster - ERROR - Unable to connect to broker ue1d-kafka-beta1a.lan.cogtree.com:9092. Continuing.
2016-07-12 20:57:07,812 [ERROR] pykafka.cluster - Unable to connect to broker ue1d-kafka-beta1a.lan.cogtree.com:9092. Continuing.
2016-07-12 20:57:07,812 - pykafka.cluster - ERROR - [Errno 111] Connection refused
Traceback (most recent call last):
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/cluster.py", line 237, in _request_metadata
    source_port=self._source_port)
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/broker.py", line 93, in __init__
    self.connect()
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/broker.py", line 202, in connect
    self._connection.connect(self._socket_timeout_ms)
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/connection.py", line 87, in connect
    (self.source_host, self.source_port)
  File "/usr/lib/python2.7/socket.py", line 571, in create_connection
    raise err
error: [Errno 111] Connection refused
2016-07-12 20:57:07,812 [ERROR] pykafka.cluster - [Errno 111] Connection refused
```

**Comment by emmett9001** [comment]
Here's [another big log](https://gist.github.com/emmett9001/b91194a6e8c97af73f51abc452b03628) from a time I was able to make the producer hang. This happened with a three-broker cluster in which I killed one of the brokers and waited a while. For the first ~5 minutes after killing the broker, the producer continued to write to the other two in the cluster without issue. Then for no apparent reason, this log happened and the producer entered a steady state of looping on `producing batch` without actually sending any messages successfully.

**Comment by emmett9001** [comment]
See #579 for another traceback that looks like it could be the same problem.

**Comment by emmett9001** [comment]
[Here](https://github.com/Parsely/pykafka/compare/2.3.1...bugfix/producer_hang) is a branch on which I've been working through some bugs we've found at Parse.ly related to producer hangs.

**Comment by nursejason** [comment]
I won't claim to know what's going on behind the scenes, but I feel as though I've been having a similar problem (On 2.3.0 and now 2.4.0). Producer is sending messages to a Kafka instance that it had been able to successfully send messages to throughout the life of the Producer (we're talking 12-24 hours), and suddenly it hangs while sending messages.

I saw this problem so often that I implemented a manual timeout using the `signal` Python module in an effort to raise an exception, terminate the Producer object, reinstantiate, and move about my business. Some Pseudocode...
```
# producer with linger_ms==50
@timeout(60) #60 == seconds until timeout
def produce(messages):
    self.producer.start()
    for message in messages:
        self.producer.produce(message) # Will eventually hang while sending messages

def shutdown():
    self.producer.stop()

#Main code block
try:
    self.producer.produce(messages) # Will eventually raise timeouterror
except TimeoutError:
    self.producer.shutdown() # Will also timeout, raising the same exception
    self.producer = new_producer() # Will never get called, because shutdown raises timeouterror

# Program will eventually raise out
# Which will print a stack trace
# But never fully exits the python application
```

It's odd because the producer hanging, even with the timeout, causes the entire object to lock up. I can start raising my exception up the call chain, but ultimately when the application goes to exit it will print the stack trace and then hang. The application never fully dies. 

I then need to manually terminate the application.
Every day.

**Comment by emmett9001** [comment]
This pull request deals with a large subset of the issues the producer can run into when a broker dies https://github.com/Parsely/pykafka/pull/595

**Comment by emmett9001** [comment]
I'm going to close this since the original issue appears to be addressed in `2.5.0.dev1`. @nursejason if you're still experiencing this bug, please open a new issue. @jox4, please let me know if your problem persists with `2.5.0.dev1`.
